## Importing the Dataset

In [1]:
import numpy as np
import pandas as pd

In [2]:
#train data 읽고 확인
train_raw = pd.read_csv("Train_data.csv")
train_raw.head()

Unnamed: 0 class  aa_000 ab_000 ac_000 ad_000 ae_000 af_000 ag_000 ag_001  \
0       52803   neg   41386     na    508    488      0      0      0      0   
1       38189   neg   29616     na   1616   1490      0      0      0      0   
2       23291   neg  241352     na     na     na     na     na      0      0   
3       16862   neg    8100     na     86     76      0      0      0      0   
4       14055   neg    2290     na    636    448      0      0      0      0   

   ...   ee_002   ee_003   ee_004  ee_005  ee_006  ee_007  ee_008 ee_009  \
0  ...   438088   202172   383094  392838  228526  104226  122526   6924   
1  ...   145524    72858   171332  308328  379466  213826    5764    292   
2  ...  3617298  2477772  3631902  997462  436380  202002  173850   1376   
3  ...    66980    36658    91898   86634   60276   23616    7518      2   
4  ...    11542     7394    14206   69592    3108     108       6      0   

  ef_000 eg_000  
0      0      0  
1      0      0  
2     na     na  
3      0      0  
4      0      0  

[5 rows x 172 columns]

In [3]:
train_raw['class'].value_counts()

neg    55934
pos     1066
Name: class, dtype: int64

In [4]:
train_raw.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 57000 entries, 0 to 56999
Columns: 172 entries, Unnamed: 0 to eg_000
dtypes: int64(2), object(170)
memory usage: 74.8+ MB


In [5]:
#data imbalance in percentage
train_raw['class'].value_counts(normalize=True)*100

neg    98.129825
pos     1.870175
Name: class, dtype: float64

In [6]:
train=train_raw.replace("na", np.nan)
train.head()

Unnamed: 0 class  aa_000 ab_000 ac_000 ad_000 ae_000 af_000 ag_000 ag_001  \
0       52803   neg   41386    NaN    508    488      0      0      0      0   
1       38189   neg   29616    NaN   1616   1490      0      0      0      0   
2       23291   neg  241352    NaN    NaN    NaN    NaN    NaN      0      0   
3       16862   neg    8100    NaN     86     76      0      0      0      0   
4       14055   neg    2290    NaN    636    448      0      0      0      0   

   ...   ee_002   ee_003   ee_004  ee_005  ee_006  ee_007  ee_008 ee_009  \
0  ...   438088   202172   383094  392838  228526  104226  122526   6924   
1  ...   145524    72858   171332  308328  379466  213826    5764    292   
2  ...  3617298  2477772  3631902  997462  436380  202002  173850   1376   
3  ...    66980    36658    91898   86634   60276   23616    7518      2   
4  ...    11542     7394    14206   69592    3108     108       6      0   

  ef_000 eg_000  
0      0      0  
1      0      0  
2    NaN    NaN  
3      0      0  
4      0      0  

[5 rows x 172 columns]

In [7]:
train.isnull().sum().sum()

810447

## Test data set from Kaggle

In [8]:
test = pd.read_csv('aps_failure_test_set.csv')
test.head()

class  aa_000 ab_000 ac_000 ad_000 ae_000 af_000 ag_000 ag_001 ag_002  ...  \
0   neg      60      0     20     12      0      0      0      0      0  ...   
1   neg      82      0     68     40      0      0      0      0      0  ...   
2   neg   66002      2    212    112      0      0      0      0      0  ...   
3   neg   59816     na   1010    936      0      0      0      0      0  ...   
4   neg    1814     na    156    140      0      0      0      0      0  ...   

   ee_002  ee_003  ee_004  ee_005   ee_006  ee_007  ee_008 ee_009 ef_000  \
0    1098     138     412     654       78      88       0      0      0   
1    1068     276    1620     116       86     462       0      0      0   
2  495076  380368  440134  269556  1315022  153680     516      0      0   
3  540820  243270  483302  485332   431376  210074  281662   3232      0   
4    7646    4144   18466   49782     3176     482      76      0      0   

  eg_000  
0      0  
1      0  
2      0  
3      0  
4      0  

[5 rows x 171 columns]

In [9]:
test['class'].value_counts()

neg    15625
pos      375
Name: class, dtype: int64

## Dataset Train-Validation Split

In [10]:
from sklearn.model_selection import train_test_split
X_train, X_val, y_train, y_val=train_test_split(train, train["class"], test_size=0.2, random_state=42, stratify=train["class"])

In [11]:
X_train.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 45600 entries, 36627 to 50505
Columns: 172 entries, Unnamed: 0 to eg_000
dtypes: int64(2), object(170)
memory usage: 60.2+ MB


In [12]:
X_train.head()

Unnamed: 0 class  aa_000 ab_000 ac_000 ad_000 ae_000 af_000 ag_000  \
36627       58126   neg   46976      0    128    124      0      0      0   
42898       72727   neg   39910    NaN     70     66      0      0      0   
23114       60535   neg   43614    NaN    152    144      0      0      0   
2962        58060   neg      60    NaN      0    NaN      0      0      0   
45204       57687   neg   38938    NaN    460    150      0      0      0   

      ag_001  ...  ee_002  ee_003  ee_004  ee_005   ee_006  ee_007  ee_008  \
36627      0  ...  176578   80224  141948  164404  1438208   32222     520   
42898      0  ...  291494  110406  265298  254714   232414  182932  309914   
23114      0  ...  314196  146948  297180  274392   247178  193972  320904   
2962       0  ...     578     190     468     732      138       0       0   
45204      0  ...  459428  220256  413674  334330   196244   92842   57548   

      ee_009 ef_000 eg_000  
36627      0      0      0  
42898    264      0      0  
23114  43452      0      0  
2962       0      0      0  
45204    268      0      0  

[5 rows x 172 columns]

In [13]:
X_train = X_train.replace(np.nan, -1) #결측치에 -1 대입
X_train01=X_train.drop(['Unnamed: 0', 'class'], axis=1)
X_train01.reset_index(drop=True, inplace=True)
X_train01

aa_000 ab_000      ac_000 ad_000 ae_000 af_000 ag_000 ag_001 ag_002  \
0       46976      0         128    124      0      0      0      0      0   
1       39910     -1          70     66      0      0      0      0      0   
2       43614     -1         152    144      0      0      0      0      0   
3          60     -1           0     -1      0      0      0      0      0   
4       38938     -1         460    150      0      0      0      0      0   
...       ...    ...         ...    ...    ...    ...    ...    ...    ...   
45595      16      0          34      6      0      0      0      0      0   
45596   42980     -1         166    128      0      0      0      0      0   
45597   37912     -1  2130706432   1530      0      0      0      0      0   
45598   33856     -1        1318   1174    192    330      0      0      0   
45599   29074     -1           0     -1      0      0      0      0      0   

      ag_003  ...  ee_002  ee_003  ee_004   ee_005   ee_006  ee_007  ee_008  \
0          0  ...  176578   80224  141948   164404  1438208   32222     520   
1          0  ...  291494  110406  265298   254714   232414  182932  309914   
2          0  ...  314196  146948  297180   274392   247178  193972  320904   
3          0  ...     578     190     468      732      138       0       0   
4          0  ...  459428  220256  413674   334330   196244   92842   57548   
...      ...  ...     ...     ...     ...      ...      ...     ...     ...   
45595      0  ...     286      68      34        0        0       0       0   
45596      0  ...  259926  134328  413012  1089582    25206   18984   14104   
45597      0  ...  800648  312142  392076   168164    63434   25862   22036   
45598      0  ...  409354  477122  142826   110344   488910    2668     160   
45599      0  ...  299858  167420  413110   332716   108034   28990    7872   

      ee_009 ef_000 eg_000  
0          0      0      0  
1        264      0      0  
2      43452      0      0  
3          0      0      0  
4        268      0      0  
...      ...    ...    ...  
45595      0      0      0  
45596   1472      0      0  
45597     24      0      0  
45598      0      0      6  
45599      0      0      0  

[45600 rows x 170 columns]

In [14]:
X_train01.dtypes

aa_000     int64
ab_000    object
ac_000    object
ad_000    object
ae_000    object
           ...  
ee_007    object
ee_008    object
ee_009    object
ef_000    object
eg_000    object
Length: 170, dtype: object

In [15]:
X_train01=X_train01.astype(float)
X_train01.dtypes

aa_000    float64
ab_000    float64
ac_000    float64
ad_000    float64
ae_000    float64
           ...   
ee_007    float64
ee_008    float64
ee_009    float64
ef_000    float64
eg_000    float64
Length: 170, dtype: object

In [16]:
y_train

36627    neg
42898    neg
23114    neg
2962     neg
45204    neg
        ... 
22448    neg
5638     neg
32020    neg
9030     neg
50505    neg
Name: class, Length: 45600, dtype: object

In [17]:
y_train.dtypes

dtype('O')

## Oversampling using SMOTE

In [18]:
from imblearn.over_sampling import SMOTE

print("Before OverSampling, counts of label '1': {}".format(sum(y_train=="pos")))
print("Before OverSampling, counts of label '0': {} \n".format(sum(y_train=="neg")))

sm = SMOTE(random_state=2)
X_train_res, y_train_res = sm.fit_sample(X_train01, y_train)

print('After OverSampling, the shape of train_X: {}'.format(X_train_res.shape))
print('After OverSampling, the shape of train_y: {} \n'.format(y_train_res.shape))

print("After OverSampling, counts of label '1': {}".format(sum(y_train_res=="pos")))
print("After OverSampling, counts of label '0': {}".format(sum(y_train_res=="neg")))

Before OverSampling, counts of label '1': 853
Before OverSampling, counts of label '0': 44747 

After OverSampling, the shape of train_X: (89494, 170)
After OverSampling, the shape of train_y: (89494,) 

After OverSampling, counts of label '1': 44747
After OverSampling, counts of label '0': 44747


## Validation set

In [19]:
X_val = X_val.replace(np.nan, -1)
X_val01=X_val.drop(['Unnamed: 0', 'class'], axis=1)
X_val01.reset_index(drop=True, inplace=True)
X_val01

aa_000 ab_000 ac_000 ad_000 ae_000 af_000 ag_000  ag_001   ag_002  \
0        39034     -1    132    108      0      0      0       0        0   
1       349286     -1     -1     -1     -1     -1      0   26204   735690   
2        32400     -1    714    626      0      0      0       0        0   
3       378224     -1     36     16      0      0      0       0        0   
4        61174     -1      0     -1      0      0      0       0        0   
...        ...    ...    ...    ...    ...    ...    ...     ...      ...   
11395    30818     -1    774    624      0      0      0       0        0   
11396    62496     -1      0     -1      0      0      0       0        0   
11397   175162     -1      0     -1      0      0      0       0        0   
11398    63624      0     20     20      0      0      0       0        0   
11399  1032510      2   7492   1102      0      0   7212  303376  2266780   

         ag_003  ...   ee_002   ee_003   ee_004    ee_005   ee_006   ee_007  \
0             0  ...   154734    69690   165178    133902   443552   691076   
1       6638176  ...  5012822  1532928  3381640   4543016   655000   207038   
2             0  ...   320846   168344   408888    330388   153806    65222   
3             0  ...    53412    24186    43278     37390    48242   293878   
4             0  ...   417030   213806   466264    560570   527686   293296   
...         ...  ...      ...      ...      ...       ...      ...      ...   
11395         0  ...   211896   112548   248956    304754   343216   141470   
11396         0  ...   709166   295914   490402    396500   268468   169826   
11397         0  ...  1639368  1118648  2449552   1915304   774144   245922   
11398         0  ...   601960   226666   554386    941358    55188    55874   
11399  12425434  ...  9383928  3999510  7595646  14600942  9247210  2594440   

       ee_008 ee_009 ef_000 eg_000  
0        6636      0      0      0  
1        3480      0     -1     -1  
2       52630     32      0      0  
3       30980    194      0      0  
4      224894   2752      0      0  
...       ...    ...    ...    ...  
11395   14940      0      0      0  
11396  334658  22344      0      0  
11397  127348     58      0      0  
11398     108      0      0      0  
11399  148896      0      0      0  

[11400 rows x 170 columns]

In [20]:
X_val01=X_val01.astype(float)
print(X_val01.dtypes)

aa_000    float64
ab_000    float64
ac_000    float64
ad_000    float64
ae_000    float64
           ...   
ee_007    float64
ee_008    float64
ee_009    float64
ef_000    float64
eg_000    float64
Length: 170, dtype: object


# Model

### RandomForest

In [21]:
from sklearn.ensemble import RandomForestClassifier
rf = RandomForestClassifier(n_jobs=4)
rf.fit(X_train_res, y_train_res)

RandomForestClassifier(n_jobs=4)

In [22]:
result = rf.predict(X_val01)
pd.Series(result).value_counts()

neg    11159
pos      241
dtype: int64

In [23]:
from sklearn.metrics import accuracy_score, f1_score
print('f1 score:', f1_score(y_val.replace({'neg': 0, 'pos' : 1}), pd.Series(result).replace({'neg': 0, 'pos' : 1})))
print('accuracy score:', accuracy_score(y_val.replace({'neg': 0, 'pos' : 1}), pd.Series(result).replace({'neg': 0, 'pos' : 1})))

f1 score: 0.7621145374449338
accuracy score: 0.9905263157894737


In [24]:
df1 = pd.DataFrame({'predicted':result, 'true':y_val})
df1.reset_index(drop=True,inplace=True)

In [25]:
i = 0
j= 0
false_neg = 0
false_pos = 0

for predicted, true in df1.values:
    if predicted != true: #예측이 틀렸을 때
        if true == 'neg': 
            i = i+10
            false_pos = false_pos+1
        else :
            j = j+500
            false_neg = false_neg+1
            
print('RandomForest 총 비용 :','$',i+j)
print('positive를 negative로 분류 :', '$',false_neg*500)
print('negative를 positive로 분류 :', '$', false_pos*10)

RandomForest 총 비용 : $ 20680
positive를 negative로 분류 : $ 20000
negative를 positive로 분류 : $ 680


### Catboost

In [26]:
from catboost import CatBoostClassifier
cat = CatBoostClassifier()
cat.fit(X_train_res, y_train_res)

Learning rate set to 0.070202
0:	learn: 0.5707008	total: 190ms	remaining: 3m 9s
1:	learn: 0.4675095	total: 299ms	remaining: 2m 29s
2:	learn: 0.3939426	total: 413ms	remaining: 2m 17s
3:	learn: 0.3311132	total: 516ms	remaining: 2m 8s
4:	learn: 0.2797918	total: 608ms	remaining: 2m 1s
5:	learn: 0.2444688	total: 711ms	remaining: 1m 57s
6:	learn: 0.2152931	total: 791ms	remaining: 1m 52s
7:	learn: 0.1924672	total: 884ms	remaining: 1m 49s
8:	learn: 0.1743603	total: 988ms	remaining: 1m 48s
9:	learn: 0.1579811	total: 1.08s	remaining: 1m 47s
10:	learn: 0.1451951	total: 1.16s	remaining: 1m 44s
11:	learn: 0.1342870	total: 1.27s	remaining: 1m 44s
12:	learn: 0.1253864	total: 1.37s	remaining: 1m 44s
13:	learn: 0.1179149	total: 1.47s	remaining: 1m 43s
14:	learn: 0.1116066	total: 1.57s	remaining: 1m 42s
15:	learn: 0.1060922	total: 1.67s	remaining: 1m 42s
16:	learn: 0.1010870	total: 1.76s	remaining: 1m 41s
17:	learn: 0.0972468	total: 1.86s	remaining: 1m 41s
18:	learn: 0.0934360	total: 1.96s	remaining: 1m

158:	learn: 0.0206768	total: 14.4s	remaining: 1m 15s
159:	learn: 0.0204996	total: 14.5s	remaining: 1m 15s
160:	learn: 0.0204319	total: 14.5s	remaining: 1m 15s
161:	learn: 0.0203024	total: 14.6s	remaining: 1m 15s
162:	learn: 0.0201139	total: 14.7s	remaining: 1m 15s
163:	learn: 0.0200554	total: 14.8s	remaining: 1m 15s
164:	learn: 0.0200179	total: 14.9s	remaining: 1m 15s
165:	learn: 0.0199508	total: 14.9s	remaining: 1m 15s
166:	learn: 0.0198886	total: 15s	remaining: 1m 15s
167:	learn: 0.0198518	total: 15.1s	remaining: 1m 14s
168:	learn: 0.0198017	total: 15.2s	remaining: 1m 14s
169:	learn: 0.0197195	total: 15.3s	remaining: 1m 14s
170:	learn: 0.0196323	total: 15.4s	remaining: 1m 14s
171:	learn: 0.0195208	total: 15.4s	remaining: 1m 14s
172:	learn: 0.0193304	total: 15.5s	remaining: 1m 14s
173:	learn: 0.0192713	total: 15.6s	remaining: 1m 14s
174:	learn: 0.0191472	total: 15.7s	remaining: 1m 14s
175:	learn: 0.0190090	total: 15.8s	remaining: 1m 14s
176:	learn: 0.0189078	total: 15.9s	remaining: 1m

317:	learn: 0.0109399	total: 28s	remaining: 60s
318:	learn: 0.0109013	total: 28.1s	remaining: 59.9s
319:	learn: 0.0108860	total: 28.1s	remaining: 59.8s
320:	learn: 0.0108446	total: 28.2s	remaining: 59.7s
321:	learn: 0.0107997	total: 28.3s	remaining: 59.6s
322:	learn: 0.0107832	total: 28.4s	remaining: 59.5s
323:	learn: 0.0107582	total: 28.5s	remaining: 59.4s
324:	learn: 0.0106848	total: 28.6s	remaining: 59.3s
325:	learn: 0.0106377	total: 28.7s	remaining: 59.2s
326:	learn: 0.0106079	total: 28.7s	remaining: 59.1s
327:	learn: 0.0105885	total: 28.8s	remaining: 59s
328:	learn: 0.0105650	total: 28.9s	remaining: 58.9s
329:	learn: 0.0105503	total: 29s	remaining: 58.8s
330:	learn: 0.0105288	total: 29.1s	remaining: 58.7s
331:	learn: 0.0105090	total: 29.1s	remaining: 58.6s
332:	learn: 0.0104976	total: 29.2s	remaining: 58.5s
333:	learn: 0.0104724	total: 29.3s	remaining: 58.4s
334:	learn: 0.0104383	total: 29.4s	remaining: 58.3s
335:	learn: 0.0103588	total: 29.5s	remaining: 58.2s
336:	learn: 0.010319

478:	learn: 0.0073815	total: 41.5s	remaining: 45.1s
479:	learn: 0.0073107	total: 41.6s	remaining: 45s
480:	learn: 0.0073107	total: 41.6s	remaining: 44.9s
481:	learn: 0.0072958	total: 41.7s	remaining: 44.8s
482:	learn: 0.0072775	total: 41.8s	remaining: 44.7s
483:	learn: 0.0072775	total: 41.9s	remaining: 44.6s
484:	learn: 0.0072670	total: 41.9s	remaining: 44.5s
485:	learn: 0.0072554	total: 42s	remaining: 44.4s
486:	learn: 0.0072348	total: 42.1s	remaining: 44.3s
487:	learn: 0.0072348	total: 42.2s	remaining: 44.2s
488:	learn: 0.0072348	total: 42.3s	remaining: 44.2s
489:	learn: 0.0072348	total: 42.3s	remaining: 44s
490:	learn: 0.0072151	total: 42.4s	remaining: 44s
491:	learn: 0.0071944	total: 42.5s	remaining: 43.9s
492:	learn: 0.0071697	total: 42.5s	remaining: 43.8s
493:	learn: 0.0071596	total: 42.6s	remaining: 43.7s
494:	learn: 0.0071596	total: 42.7s	remaining: 43.6s
495:	learn: 0.0071311	total: 42.8s	remaining: 43.5s
496:	learn: 0.0071263	total: 42.9s	remaining: 43.4s
497:	learn: 0.007110

640:	learn: 0.0066329	total: 53.3s	remaining: 29.8s
641:	learn: 0.0066329	total: 53.3s	remaining: 29.7s
642:	learn: 0.0066329	total: 53.4s	remaining: 29.7s
643:	learn: 0.0066329	total: 53.5s	remaining: 29.6s
644:	learn: 0.0066329	total: 53.6s	remaining: 29.5s
645:	learn: 0.0066328	total: 53.6s	remaining: 29.4s
646:	learn: 0.0066328	total: 53.7s	remaining: 29.3s
647:	learn: 0.0066328	total: 53.8s	remaining: 29.2s
648:	learn: 0.0066328	total: 53.8s	remaining: 29.1s
649:	learn: 0.0066328	total: 53.9s	remaining: 29s
650:	learn: 0.0066328	total: 54s	remaining: 28.9s
651:	learn: 0.0066328	total: 54s	remaining: 28.8s
652:	learn: 0.0066328	total: 54.1s	remaining: 28.8s
653:	learn: 0.0066328	total: 54.2s	remaining: 28.7s
654:	learn: 0.0066221	total: 54.2s	remaining: 28.6s
655:	learn: 0.0066221	total: 54.3s	remaining: 28.5s
656:	learn: 0.0066220	total: 54.4s	remaining: 28.4s
657:	learn: 0.0066220	total: 54.4s	remaining: 28.3s
658:	learn: 0.0066220	total: 54.5s	remaining: 28.2s
659:	learn: 0.0066

801:	learn: 0.0064850	total: 1m 4s	remaining: 16s
802:	learn: 0.0064849	total: 1m 4s	remaining: 15.9s
803:	learn: 0.0064849	total: 1m 4s	remaining: 15.8s
804:	learn: 0.0064849	total: 1m 4s	remaining: 15.7s
805:	learn: 0.0064849	total: 1m 4s	remaining: 15.6s
806:	learn: 0.0064849	total: 1m 4s	remaining: 15.5s
807:	learn: 0.0064848	total: 1m 5s	remaining: 15.5s
808:	learn: 0.0064848	total: 1m 5s	remaining: 15.4s
809:	learn: 0.0064848	total: 1m 5s	remaining: 15.3s
810:	learn: 0.0064848	total: 1m 5s	remaining: 15.2s
811:	learn: 0.0064848	total: 1m 5s	remaining: 15.1s
812:	learn: 0.0064848	total: 1m 5s	remaining: 15.1s
813:	learn: 0.0064848	total: 1m 5s	remaining: 15s
814:	learn: 0.0064848	total: 1m 5s	remaining: 14.9s
815:	learn: 0.0064847	total: 1m 5s	remaining: 14.8s
816:	learn: 0.0064847	total: 1m 5s	remaining: 14.7s
817:	learn: 0.0064847	total: 1m 5s	remaining: 14.6s
818:	learn: 0.0064847	total: 1m 5s	remaining: 14.6s
819:	learn: 0.0064847	total: 1m 5s	remaining: 14.5s
820:	learn: 0.00

960:	learn: 0.0064749	total: 1m 17s	remaining: 3.13s
961:	learn: 0.0064749	total: 1m 17s	remaining: 3.04s
962:	learn: 0.0064749	total: 1m 17s	remaining: 2.96s
963:	learn: 0.0064749	total: 1m 17s	remaining: 2.88s
964:	learn: 0.0064749	total: 1m 17s	remaining: 2.8s
965:	learn: 0.0064749	total: 1m 17s	remaining: 2.72s
966:	learn: 0.0064749	total: 1m 17s	remaining: 2.64s
967:	learn: 0.0064749	total: 1m 17s	remaining: 2.56s
968:	learn: 0.0064749	total: 1m 17s	remaining: 2.48s
969:	learn: 0.0064749	total: 1m 17s	remaining: 2.4s
970:	learn: 0.0064749	total: 1m 17s	remaining: 2.32s
971:	learn: 0.0064749	total: 1m 17s	remaining: 2.24s
972:	learn: 0.0064749	total: 1m 17s	remaining: 2.16s
973:	learn: 0.0064749	total: 1m 18s	remaining: 2.08s
974:	learn: 0.0064749	total: 1m 18s	remaining: 2s
975:	learn: 0.0064749	total: 1m 18s	remaining: 1.92s
976:	learn: 0.0064749	total: 1m 18s	remaining: 1.84s
977:	learn: 0.0064749	total: 1m 18s	remaining: 1.76s
978:	learn: 0.0064749	total: 1m 18s	remaining: 1.68

In [27]:
result2 = cat.predict(X_val01)
pd.Series(result2).value_counts()

neg    11159
pos      241
dtype: int64

In [28]:
print('f1 score:', f1_score(y_val.replace({'neg': 0, 'pos' : 1}), pd.Series(result2).replace({'neg': 0, 'pos' : 1})))
print('accuracy score:', accuracy_score(y_val.replace({'neg': 0, 'pos' : 1}), pd.Series(result2).replace({'neg': 0, 'pos' : 1})))

f1 score: 0.7929515418502203
accuracy score: 0.9917543859649123


In [29]:
df2 = pd.DataFrame({'predicted':result2, 'true':y_val})
df2.reset_index(drop=True,inplace=True)

In [30]:
i = 0
j= 0
false_neg = 0
false_pos = 0

for predicted, true in df2.values:
    if predicted != true: #예측이 틀렸을 때
        if true == 'neg': 
            i = i+10
            false_pos = false_pos+1
        else :
            j = j+500
            false_neg = false_neg+1
            
print('CatBoost 총 비용 :','$',i+j)
print('positive를 negative로 분류 :', '$',false_neg*500)
print('negative를 positive로 분류 :', '$', false_pos*10)

CatBoost 총 비용 : $ 17110
positive를 negative로 분류 : $ 16500
negative를 positive로 분류 : $ 610


### XGBoost

In [31]:
import xgboost as xgb
xgb_model = xgb.XGBClassifier(objective="binary:logistic", random_state=42)
xgb_model.fit(X_train_res, y_train_res)

XGBClassifier(base_score=0.5, booster='gbtree', colsample_bylevel=1,
              colsample_bynode=1, colsample_bytree=1, gamma=0, gpu_id=-1,
              importance_type='gain', interaction_constraints='',
              learning_rate=0.300000012, max_delta_step=0, max_depth=6,
              min_child_weight=1, missing=nan, monotone_constraints='()',
              n_estimators=100, n_jobs=0, num_parallel_tree=1, random_state=42,
              reg_alpha=0, reg_lambda=1, scale_pos_weight=1, subsample=1,
              tree_method='exact', validate_parameters=1, verbosity=None)

In [32]:
result3 = xgb_model.predict(X_val01)
pd.Series(result3).value_counts()

neg    11175
pos      225
dtype: int64

In [33]:
print('f1 score:', f1_score(y_val.replace({'neg': 0, 'pos' : 1}), pd.Series(result3).replace({'neg': 0, 'pos' : 1})))
print('accuracy score:', accuracy_score(y_val.replace({'neg': 0, 'pos' : 1}), pd.Series(result3).replace({'neg': 0, 'pos' : 1})))

f1 score: 0.817351598173516
accuracy score: 0.9929824561403509


In [34]:
df3 = pd.DataFrame({'predicted':result3, 'true':y_val})
df3.reset_index(drop=True,inplace=True)

In [35]:
i = 0
j= 0
false_neg = 0
false_pos = 0

for predicted, true in df3.values:
    if predicted != true: #예측이 틀렸을 때
        if true == 'neg': 
            i = i+10
            false_pos = false_pos+1
        else :
            j = j+500
            false_neg = false_neg+1
            
print('XGBoost 총 비용 :','$',i+j)
print('positive를 negative로 분류 :',false_neg,'개','$',false_neg*500)
print('negative를 positive로 분류 :', false_pos,'개', '$', false_pos*10)

XGBoost 총 비용 : $ 17460
positive를 negative로 분류 : 34 개 $ 17000
negative를 positive로 분류 : 46 개 $ 460


### Logistic Regression

In [36]:
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import f1_score, log_loss, confusion_matrix,classification_report
logistic_clf= LogisticRegression(random_state=0, solver='lbfgs',C=1.0).fit(X_train_res, y_train_res)

C:\Users\LeeYoungsin\AppData\Roaming\Python\Python37\site-packages\sklearn\linear_model\_logistic.py:764: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  extra_warning_msg=_LOGISTIC_SOLVER_CONVERGENCE_MSG)


In [37]:
result4= logistic_clf.predict(X_val01)
pd.Series(result4).value_counts()

neg    10762
pos      638
dtype: int64

In [38]:
print('f1 score:', f1_score(y_val.replace({'neg': 0, 'pos' : 1}), pd.Series(result4).replace({'neg': 0, 'pos' : 1})))
print('accuracy score:', accuracy_score(y_val.replace({'neg': 0, 'pos' : 1}), pd.Series(result4).replace({'neg': 0, 'pos' : 1})))

f1 score: 0.45593419506462984
accuracy score: 0.9593859649122807


In [39]:
df4 = pd.DataFrame({'predicted':result4, 'true':y_val})
df4.reset_index(drop=True,inplace=True)

In [40]:
i = 0
j= 0
false_neg = 0
false_pos = 0

for predicted, true in df4.values:
    if predicted != true: #예측이 틀렸을 때
        if true == 'neg': 
            i = i+10
            false_pos = false_pos+1
        else :
            j = j+500
            false_neg = false_neg+1
            
print('Logistic Regression 총 비용 :','$',i+j)
print('positive를 negative로 분류 :', '$',false_neg*500)
print('negative를 positive로 분류 :', '$', false_pos*10)

Logistic Regression 총 비용 : $ 13940
positive를 negative로 분류 : $ 9500
negative를 positive로 분류 : $ 4440


### Softmax Regression

In [41]:
softmax_reg = LogisticRegression(multi_class="multinomial", solver="lbfgs", C=10)
softmax_reg.fit(X_train_res, y_train_res)

C:\Users\LeeYoungsin\AppData\Roaming\Python\Python37\site-packages\sklearn\linear_model\_logistic.py:764: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  extra_warning_msg=_LOGISTIC_SOLVER_CONVERGENCE_MSG)


LogisticRegression(C=10, multi_class='multinomial')

In [42]:
result5= softmax_reg.predict(X_val01)
pd.Series(result5).value_counts()

neg    10766
pos      634
dtype: int64

In [43]:
print('f1 score:', f1_score(y_val.replace({'neg': 0, 'pos' : 1}), pd.Series(result5).replace({'neg': 0, 'pos' : 1})))
print('accuracy score:', accuracy_score(y_val.replace({'neg': 0, 'pos' : 1}), pd.Series(result5).replace({'neg': 0, 'pos' : 1})))

f1 score: 0.45808736717827625
accuracy score: 0.9597368421052631


In [44]:
df5 = pd.DataFrame({'predicted':result5, 'true':y_val})
df5.reset_index(drop=True,inplace=True)

In [45]:
i = 0
j= 0
false_neg = 0
false_pos = 0

for predicted, true in df5.values:
    if predicted != true: #예측이 틀렸을 때
        if true == 'neg': 
            i = i+10
            false_pos = false_pos+1
        else :
            j = j+500
            false_neg = false_neg+1
            
print('SoftMax 총 비용 :','$',i+j)
print('positive를 negative로 분류 :', '$',false_neg*500)
print('negative를 positive로 분류 :', '$', false_pos*10)

SoftMax 총 비용 : $ 13900
positive를 negative로 분류 : $ 9500
negative를 positive로 분류 : $ 4400


### Linear SVC

In [46]:
from sklearn.pipeline import Pipeline
from sklearn.preprocessing import StandardScaler
from sklearn.model_selection import GridSearchCV
from sklearn.metrics import confusion_matrix
from sklearn.svm import LinearSVC
from sklearn.svm import SVC

In [47]:
Linear_svm_clf =Pipeline((
    ("scaler", StandardScaler()),
    ("linear_svc", LinearSVC(C=0.1, loss="hinge")),
))                                                     
Linear_svm_clf.fit(X_train_res, y_train_res)

C:\Users\LeeYoungsin\AppData\Roaming\Python\Python37\site-packages\sklearn\svm\_base.py:977: ConvergenceWarning: Liblinear failed to converge, increase the number of iterations.
  "the number of iterations.", ConvergenceWarning)


Pipeline(steps=[('scaler', StandardScaler()),
                ('linear_svc', LinearSVC(C=0.1, loss='hinge'))])

In [48]:
result6=Linear_svm_clf.predict(X_val01)
pd.Series(result6).value_counts()

neg    10945
pos      455
dtype: int64

In [49]:
print('f1 score:', f1_score(y_val.replace({'neg': 0, 'pos' : 1}), pd.Series(result6).replace({'neg': 0, 'pos' : 1})))
print('accuracy score:', accuracy_score(y_val.replace({'neg': 0, 'pos' : 1}), pd.Series(result6).replace({'neg': 0, 'pos' : 1})))

f1 score: 0.5898203592814372
accuracy score: 0.9759649122807017


In [50]:
df6 = pd.DataFrame({'predicted':result4, 'true':y_val})
df6.reset_index(drop=True,inplace=True)

In [51]:
i = 0
j= 0
false_neg = 0
false_pos = 0

for predicted, true in df6.values:
    if predicted != true: #예측이 틀렸을 때
        if true == 'neg': 
            i = i+10
            false_pos = false_pos+1
        else :
            j = j+500
            false_neg = false_neg+1
            
print('LinearSVC 총 비용 :','$',i+j)
print('positive를 negative로 분류 :', '$',false_neg*500)
print('negative를 positive로 분류 :', '$', false_pos*10)

LinearSVC 총 비용 : $ 13940
positive를 negative로 분류 : $ 9500
negative를 positive로 분류 : $ 4440


## Testing the Model

In [52]:
test = test.replace("na", -1)
test1 = test.drop(['class'], axis=1)
test1

aa_000 ab_000      ac_000 ad_000 ae_000 af_000 ag_000 ag_001 ag_002  \
0          60      0          20     12      0      0      0      0      0   
1          82      0          68     40      0      0      0      0      0   
2       66002      2         212    112      0      0      0      0      0   
3       59816     -1        1010    936      0      0      0      0      0   
4        1814     -1         156    140      0      0      0      0      0   
...       ...    ...         ...    ...    ...    ...    ...    ...    ...   
15995   81852     -1  2130706432    892      0      0      0      0      0   
15996      18      0          52     46      8     26      0      0      0   
15997   79636     -1        1670   1518      0      0      0      0      0   
15998     110     -1          36     32      0      0      0      0      0   
15999       8      0           6      4      2      2      0      0      0   

       ag_003  ...  ee_002  ee_003  ee_004  ee_005   ee_006  ee_007  ee_008  \
0        2682  ...    1098     138     412     654       78      88       0   
1           0  ...    1068     276    1620     116       86     462       0   
2      199486  ...  495076  380368  440134  269556  1315022  153680     516   
3           0  ...  540820  243270  483302  485332   431376  210074  281662   
4           0  ...    7646    4144   18466   49782     3176     482      76   
...       ...  ...     ...     ...     ...     ...      ...     ...     ...   
15995       0  ...  632658  273242  510354  373918   349840  317840  960024   
15996       0  ...     266      44      46      14        2       0       0   
15997       0  ...  806832  449962  778826  581558   375498  222866  358934   
15998       0  ...     588     210     180     544     1004    1338      74   
15999       0  ...      46      10      48      14       42      46       0   

      ee_009 ef_000 eg_000  
0          0      0      0  
1          0      0      0  
2          0      0      0  
3       3232      0      0  
4          0      0      0  
...      ...    ...    ...  
15995  25566      0      0  
15996      0      0      0  
15997  19548      0      0  
15998      0      0      0  
15999      0      0      0  

[16000 rows x 170 columns]

In [53]:
test1=test1.astype(float)
test1.dtypes

aa_000    float64
ab_000    float64
ac_000    float64
ad_000    float64
ae_000    float64
           ...   
ee_007    float64
ee_008    float64
ee_009    float64
ef_000    float64
eg_000    float64
Length: 170, dtype: object

### RandomForest

In [54]:
test_pred1 = rf.predict(test1)
pd.Series(test_pred1).value_counts()

neg    15618
pos      382
dtype: int64

In [55]:
print('f1 score:', f1_score(test['class'].replace({'neg': 0, 'pos' : 1}), pd.Series(test_pred1).replace({'neg': 0, 'pos' : 1})))
print('accuracy score:', accuracy_score(test['class'].replace({'neg': 0, 'pos' : 1}), pd.Series(test_pred1).replace({'neg': 0, 'pos' : 1})))

f1 score: 0.9299867899603699
accuracy score: 0.9966875


In [56]:
t_df1 = pd.DataFrame({'predicted':test_pred1, 'true':test['class']})
t_df1.reset_index(drop=True,inplace=True)

In [57]:
i = 0
j= 0
false_neg = 0
false_pos = 0

for predicted, true in t_df1.values:
    if predicted != true: #예측이 틀렸을 때
        if true == 'neg': 
            i = i+10
            false_pos = false_pos+1
        else :
            j = j+500
            false_neg = false_neg+1
            
print('RandomForest 총 비용 :','$',i+j)
print('positive를 negative로 분류 :', '$',false_neg*500)
print('negative를 positive로 분류 :', '$', false_pos*10)

RandomForest 총 비용 : $ 11800
positive를 negative로 분류 : $ 11500
negative를 positive로 분류 : $ 300


### Catboost

In [58]:
test_pred2 = cat.predict(test1)
pd.Series(test_pred2).value_counts()

neg    15612
pos      388
dtype: int64

In [59]:
print('f1 score:', f1_score(test['class'].replace({'neg': 0, 'pos' : 1}), pd.Series(test_pred2).replace({'neg': 0, 'pos' : 1})))
print('accuracy score:', accuracy_score(test['class'].replace({'neg': 0, 'pos' : 1}), pd.Series(test_pred2).replace({'neg': 0, 'pos' : 1})))

f1 score: 0.9226736566186108
accuracy score: 0.9963125


In [60]:
t_df2 = pd.DataFrame({'predicted':test_pred2, 'true':test['class']})
t_df2.reset_index(drop=True,inplace=True)

In [61]:
i = 0
j= 0
false_neg = 0
false_pos = 0

for predicted, true in t_df2.values:
    if predicted != true: #예측이 틀렸을 때
        if true == 'neg': 
            i = i+10
            false_pos = false_pos+1
        else :
            j = j+500
            false_neg = false_neg+1
            
print('CatBoost 총 비용 :','$',i+j)
print('positive를 negative로 분류 :', '$',false_neg*500)
print('negative를 positive로 분류 :', '$', false_pos*10)

CatBoost 총 비용 : $ 11860
positive를 negative로 분류 : $ 11500
negative를 positive로 분류 : $ 360


### XGBoost

In [62]:
test_pred3 = xgb_model.predict(test1)
pd.Series(test_pred3).value_counts()

neg    15628
pos      372
dtype: int64

In [63]:
print('f1 score:', f1_score(test['class'].replace({'neg': 0, 'pos' : 1}), pd.Series(test_pred3).replace({'neg': 0, 'pos' : 1})))
print('accuracy score:', accuracy_score(test['class'].replace({'neg': 0, 'pos' : 1}), pd.Series(test_pred3).replace({'neg': 0, 'pos' : 1})))

f1 score: 0.9477911646586344
accuracy score: 0.9975625


In [64]:
t_df3 = pd.DataFrame({'predicted':test_pred3, 'true':test['class']})
t_df3.reset_index(drop=True,inplace=True)

In [65]:
i = 0
j= 0
false_neg = 0
false_pos = 0

for predicted, true in t_df3.values:
    if predicted != true: #예측이 틀렸을 때
        if true == 'neg': 
            i = i+10
            false_pos = false_pos+1
        else :
            j = j+500
            false_neg = false_neg+1
            
print('XG Boost 총 비용 :','$',i+j)
print('positive를 negative로 분류 :', '$',false_neg*500)
print('negative를 positive로 분류 :', '$', false_pos*10)

XG Boost 총 비용 : $ 10680
positive를 negative로 분류 : $ 10500
negative를 positive로 분류 : $ 180


### Logistic Regression

In [66]:
test_pred4 = logistic_clf.predict(test1)
pd.Series(test_pred4).value_counts()

neg    15061
pos      939
dtype: int64

In [67]:
print('f1 score:', f1_score(test['class'].replace({'neg': 0, 'pos' : 1}), pd.Series(test_pred4).replace({'neg': 0, 'pos' : 1})))
print('accuracy score:', accuracy_score(test['class'].replace({'neg': 0, 'pos' : 1}), pd.Series(test_pred4).replace({'neg': 0, 'pos' : 1})))

f1 score: 0.5144596651445965
accuracy score: 0.960125


In [68]:
t_df4 = pd.DataFrame({'predicted':test_pred4, 'true':test['class']})
t_df4.reset_index(drop=True,inplace=True)

In [69]:
i = 0
j= 0
false_neg = 0
false_pos = 0

for predicted, true in t_df4.values:
    if predicted != true: #예측이 틀렸을 때
        if true == 'neg': 
            i = i+10
            false_pos = false_pos+1
        else :
            j = j+500
            false_neg = false_neg+1
            
print('Logistic Regression 총 비용 :','$',i+j)
print('positive를 negative로 분류 :', '$',false_neg*500)
print('negative를 positive로 분류 :', '$', false_pos*10)

Logistic Regression 총 비용 : $ 24510
positive를 negative로 분류 : $ 18500
negative를 positive로 분류 : $ 6010


### Softmax Regression

In [70]:
test_pred5 = softmax_reg.predict(test1)
pd.Series(test_pred5).value_counts()

neg    15047
pos      953
dtype: int64

In [71]:
print('f1 score:', f1_score(test['class'].replace({'neg': 0, 'pos' : 1}), pd.Series(test_pred5).replace({'neg': 0, 'pos' : 1})))
print('accuracy score:', accuracy_score(test['class'].replace({'neg': 0, 'pos' : 1}), pd.Series(test_pred5).replace({'neg': 0, 'pos' : 1})))

f1 score: 0.5150602409638554
accuracy score: 0.95975


In [72]:
t_df5 = pd.DataFrame({'predicted':test_pred5, 'true':test['class']})
t_df5.reset_index(drop=True,inplace=True)

In [73]:
i = 0
j= 0
false_neg = 0
false_pos = 0

for predicted, true in t_df5.values:
    if predicted != true: #예측이 틀렸을 때
        if true == 'neg': 
            i = i+10
            false_pos = false_pos+1
        else :
            j = j+500
            false_neg = false_neg+1
            
print('SoftMax 총 비용 :','$',i+j)
print('positive를 negative로 분류 :', '$',false_neg*500)
print('negative를 positive로 분류 :', '$', false_pos*10)

SoftMax 총 비용 : $ 22610
positive를 negative로 분류 : $ 16500
negative를 positive로 분류 : $ 6110


### Linear SVC

In [74]:
test_pred6=Linear_svm_clf.predict(test1)
pd.Series(test_pred6).value_counts()

neg    15318
pos      682
dtype: int64

In [75]:
print('f1 score:', f1_score(test['class'].replace({'neg': 0, 'pos' : 1}), pd.Series(test_pred6).replace({'neg': 0, 'pos' : 1})))
print('accuracy score:', accuracy_score(test['class'].replace({'neg': 0, 'pos' : 1}), pd.Series(test_pred6).replace({'neg': 0, 'pos' : 1})))

f1 score: 0.664143803216651
accuracy score: 0.9778125


In [76]:
t_df6 = pd.DataFrame({'predicted':test_pred6, 'true':test['class']})
t_df6.reset_index(drop=True,inplace=True)

In [77]:
i = 0
j= 0
false_neg = 0
false_pos = 0

for predicted, true in t_df6.values:
    if predicted != true: #예측이 틀렸을 때
        if true == 'neg': 
            i = i+10
            false_pos = false_pos+1
        else :
            j = j+500
            false_neg = false_neg+1
            
print('LinearSVC 총 비용 :','$',i+j)
print('positive를 negative로 분류 :', '$',false_neg*500)
print('negative를 positive로 분류 :', '$', false_pos*10)

LinearSVC 총 비용 : $ 15310
positive를 negative로 분류 : $ 12000
negative를 positive로 분류 : $ 3310


## Additional Parameter Tuning on XGB and CB

### insight

The weight for class 1 in binary classification. The value is used as a multiplier for the weights of objects from class 1. <br>
<https://catboost.ai/docs/concepts/python-reference_parameters-list.html#python-reference_parameters-list>

Tip. For imbalanced datasets, the weight multiplier can be set to $\frac{sum-negative}{sum-positive}$ <br>
oversampling 하기 전 y_train(class label)의 pos:neg=1:52 

**1. XGB Params Tuning**

In [78]:
import xgboost as xgb
xgb_model_2 = xgb.XGBClassifier(learning_rate=0.25, subsample=0.8, objective="binary:logistic", random_state=42)
xgb_model_2.fit(X_train_res, y_train_res)

XGBClassifier(base_score=0.5, booster='gbtree', colsample_bylevel=1,
              colsample_bynode=1, colsample_bytree=1, gamma=0, gpu_id=-1,
              importance_type='gain', interaction_constraints='',
              learning_rate=0.25, max_delta_step=0, max_depth=6,
              min_child_weight=1, missing=nan, monotone_constraints='()',
              n_estimators=100, n_jobs=0, num_parallel_tree=1, random_state=42,
              reg_alpha=0, reg_lambda=1, scale_pos_weight=1, subsample=0.8,
              tree_method='exact', validate_parameters=1, verbosity=None)

In [79]:
result7 = xgb_model_2.predict(X_val01)
pd.Series(result7).value_counts()

neg    11171
pos      229
dtype: int64

In [80]:
print('f1 score:', f1_score(y_val.replace({'neg': 0, 'pos' : 1}), pd.Series(result7).replace({'neg': 0, 'pos' : 1})))
print('accuracy score:', accuracy_score(y_val.replace({'neg': 0, 'pos' : 1}), pd.Series(result7).replace({'neg': 0, 'pos' : 1})))

f1 score: 0.828054298642534
accuracy score: 0.9933333333333333


In [81]:
df7 = pd.DataFrame({'predicted':result7, 'true':y_val})
df7.reset_index(drop=True,inplace=True)

In [82]:
i = 0
j= 0
false_neg = 0
false_pos = 0

for predicted, true in df7.values:
    if predicted != true: #예측이 틀렸을 때
        if true == 'neg': 
            i = i+10
            false_pos = false_pos+1
        else :
            j = j+500
            false_neg = false_neg+1
            
print('XGBoost 총 비용 :','$',i+j)
print('positive를 negative로 분류 :', '$',false_neg,'개',false_neg*500)
print('negative를 positive로 분류 :', '$',false_pos,'개', false_pos*10)

XGBoost 총 비용 : $ 15460
positive를 negative로 분류 : $ 30 개 15000
negative를 positive로 분류 : $ 46 개 460


#### 1-1 XGB scale_pos_weight setting to 52

In [83]:
import xgboost as xgb
xgb_model_3 = xgb.XGBClassifier(learning_rate=0.2, subsample=0.8, objective="binary:logistic", scale_pos_weight=52, random_state=42)
xgb_model_3.fit(X_train_res, y_train_res)

XGBClassifier(base_score=0.5, booster='gbtree', colsample_bylevel=1,
              colsample_bynode=1, colsample_bytree=1, gamma=0, gpu_id=-1,
              importance_type='gain', interaction_constraints='',
              learning_rate=0.2, max_delta_step=0, max_depth=6,
              min_child_weight=1, missing=nan, monotone_constraints='()',
              n_estimators=100, n_jobs=0, num_parallel_tree=1, random_state=42,
              reg_alpha=0, reg_lambda=1, scale_pos_weight=52, subsample=0.8,
              tree_method='exact', validate_parameters=1, verbosity=None)

In [84]:
result9 = xgb_model_3.predict(X_val01)
pd.Series(result9).value_counts()

neg    11094
pos      306
dtype: int64

In [85]:
print('f1 score:', f1_score(y_val.replace({'neg': 0, 'pos' : 1}), pd.Series(result9).replace({'neg': 0, 'pos' : 1})))
print('accuracy score:', accuracy_score(y_val.replace({'neg': 0, 'pos' : 1}), pd.Series(result9).replace({'neg': 0, 'pos' : 1})))

f1 score: 0.7321772639691715
accuracy score: 0.9878070175438597


In [86]:
df9 = pd.DataFrame({'predicted':result9, 'true':y_val})
df9.reset_index(drop=True,inplace=True)

In [87]:
i = 0
j= 0
false_neg = 0
false_pos = 0

for predicted, true in df9.values:
    if predicted != true: #예측이 틀렸을 때
        if true == 'neg': 
            i = i+10
            false_pos = false_pos+1
        else :
            j = j+500
            false_neg = false_neg+1
            
print('XGBoost 총 비용 :','$',i+j)
print('positive를 negative로 분류 :', '$',false_neg,'개',false_neg*500)
print('negative를 positive로 분류 :', '$',false_pos,'개', false_pos*10)

XGBoost 총 비용 : $ 12660
positive를 negative로 분류 : $ 23 개 11500
negative를 positive로 분류 : $ 116 개 1160


scale_pos_weight 을 52에 가까이 할수록 총 비용은 줄어들지만
f1 score은 점점 낮아진다. <br>
모델의 목적이 false_negative 분류에 의한 비용 감소니까, trade-off에서 비용을 감소시키는 쪽을 택한다. <br>
모델의 하이퍼파라미터 튜닝을 할 때 무엇이 모델의 목적에 우선순위인지 파악해야 한다.

**2. CatBoost Params Tuning**

기존 learning rate = 0.07<br>
Setting learning rate to 0.3 and scale_pos_weight= 10

In [88]:
from catboost import CatBoostClassifier
cat2 = CatBoostClassifier(learning_rate=0.3, scale_pos_weight=10, verbose=True)
cat2.fit(X_train_res, y_train_res)

0:	learn: 0.2358208	total: 117ms	remaining: 1m 57s
1:	learn: 0.1087975	total: 214ms	remaining: 1m 46s
2:	learn: 0.0705597	total: 311ms	remaining: 1m 43s
3:	learn: 0.0568091	total: 425ms	remaining: 1m 45s
4:	learn: 0.0473283	total: 559ms	remaining: 1m 51s
5:	learn: 0.0425090	total: 650ms	remaining: 1m 47s
6:	learn: 0.0372743	total: 743ms	remaining: 1m 45s
7:	learn: 0.0348986	total: 829ms	remaining: 1m 42s
8:	learn: 0.0314994	total: 926ms	remaining: 1m 41s
9:	learn: 0.0301612	total: 1.02s	remaining: 1m 41s
10:	learn: 0.0281558	total: 1.11s	remaining: 1m 40s
11:	learn: 0.0271557	total: 1.2s	remaining: 1m 38s
12:	learn: 0.0257247	total: 1.29s	remaining: 1m 37s
13:	learn: 0.0247208	total: 1.39s	remaining: 1m 37s
14:	learn: 0.0231230	total: 1.48s	remaining: 1m 37s
15:	learn: 0.0222602	total: 1.58s	remaining: 1m 37s
16:	learn: 0.0211968	total: 1.68s	remaining: 1m 37s
17:	learn: 0.0203771	total: 1.79s	remaining: 1m 37s
18:	learn: 0.0195659	total: 1.9s	remaining: 1m 37s
19:	learn: 0.0189183	tot

158:	learn: 0.0107876	total: 13s	remaining: 1m 8s
159:	learn: 0.0107875	total: 13.1s	remaining: 1m 8s
160:	learn: 0.0107875	total: 13.2s	remaining: 1m 8s
161:	learn: 0.0107873	total: 13.3s	remaining: 1m 8s
162:	learn: 0.0107873	total: 13.3s	remaining: 1m 8s
163:	learn: 0.0107873	total: 13.4s	remaining: 1m 8s
164:	learn: 0.0107873	total: 13.5s	remaining: 1m 8s
165:	learn: 0.0107873	total: 13.6s	remaining: 1m 8s
166:	learn: 0.0107873	total: 13.6s	remaining: 1m 7s
167:	learn: 0.0107873	total: 13.7s	remaining: 1m 7s
168:	learn: 0.0107873	total: 13.8s	remaining: 1m 7s
169:	learn: 0.0107873	total: 13.8s	remaining: 1m 7s
170:	learn: 0.0107873	total: 13.9s	remaining: 1m 7s
171:	learn: 0.0107873	total: 14s	remaining: 1m 7s
172:	learn: 0.0107873	total: 14.1s	remaining: 1m 7s
173:	learn: 0.0107873	total: 14.1s	remaining: 1m 7s
174:	learn: 0.0107873	total: 14.2s	remaining: 1m 6s
175:	learn: 0.0107873	total: 14.3s	remaining: 1m 6s
176:	learn: 0.0107873	total: 14.3s	remaining: 1m 6s
177:	learn: 0.01

318:	learn: 0.0086080	total: 24.9s	remaining: 53.2s
319:	learn: 0.0086079	total: 25s	remaining: 53.1s
320:	learn: 0.0086079	total: 25s	remaining: 53s
321:	learn: 0.0086079	total: 25.1s	remaining: 52.9s
322:	learn: 0.0086077	total: 25.2s	remaining: 52.8s
323:	learn: 0.0086076	total: 25.2s	remaining: 52.7s
324:	learn: 0.0086076	total: 25.3s	remaining: 52.6s
325:	learn: 0.0086076	total: 25.4s	remaining: 52.5s
326:	learn: 0.0086074	total: 25.4s	remaining: 52.4s
327:	learn: 0.0086073	total: 25.5s	remaining: 52.3s
328:	learn: 0.0086073	total: 25.6s	remaining: 52.2s
329:	learn: 0.0086069	total: 25.7s	remaining: 52.1s
330:	learn: 0.0086068	total: 25.7s	remaining: 52s
331:	learn: 0.0086068	total: 25.8s	remaining: 51.9s
332:	learn: 0.0086066	total: 25.9s	remaining: 51.8s
333:	learn: 0.0086065	total: 25.9s	remaining: 51.7s
334:	learn: 0.0086064	total: 26s	remaining: 51.6s
335:	learn: 0.0086063	total: 26.1s	remaining: 51.5s
336:	learn: 0.0086062	total: 26.2s	remaining: 51.5s
337:	learn: 0.0085800	

479:	learn: 0.0083314	total: 36.1s	remaining: 39.1s
480:	learn: 0.0083314	total: 36.2s	remaining: 39s
481:	learn: 0.0083314	total: 36.2s	remaining: 38.9s
482:	learn: 0.0083314	total: 36.3s	remaining: 38.9s
483:	learn: 0.0083314	total: 36.4s	remaining: 38.8s
484:	learn: 0.0083314	total: 36.5s	remaining: 38.7s
485:	learn: 0.0083314	total: 36.5s	remaining: 38.6s
486:	learn: 0.0082337	total: 36.6s	remaining: 38.6s
487:	learn: 0.0082336	total: 36.7s	remaining: 38.5s
488:	learn: 0.0082335	total: 36.8s	remaining: 38.4s
489:	learn: 0.0082335	total: 36.8s	remaining: 38.3s
490:	learn: 0.0082335	total: 36.9s	remaining: 38.2s
491:	learn: 0.0082335	total: 37s	remaining: 38.2s
492:	learn: 0.0082335	total: 37s	remaining: 38.1s
493:	learn: 0.0082335	total: 37.1s	remaining: 38s
494:	learn: 0.0082335	total: 37.2s	remaining: 37.9s
495:	learn: 0.0082335	total: 37.2s	remaining: 37.8s
496:	learn: 0.0082335	total: 37.3s	remaining: 37.8s
497:	learn: 0.0082335	total: 37.4s	remaining: 37.7s
498:	learn: 0.008233

640:	learn: 0.0073333	total: 47.9s	remaining: 26.8s
641:	learn: 0.0073333	total: 48s	remaining: 26.8s
642:	learn: 0.0073333	total: 48s	remaining: 26.7s
643:	learn: 0.0073333	total: 48.1s	remaining: 26.6s
644:	learn: 0.0073333	total: 48.2s	remaining: 26.5s
645:	learn: 0.0073333	total: 48.3s	remaining: 26.4s
646:	learn: 0.0073333	total: 48.3s	remaining: 26.4s
647:	learn: 0.0073333	total: 48.4s	remaining: 26.3s
648:	learn: 0.0073333	total: 48.5s	remaining: 26.2s
649:	learn: 0.0073333	total: 48.5s	remaining: 26.1s
650:	learn: 0.0073333	total: 48.6s	remaining: 26.1s
651:	learn: 0.0073333	total: 48.7s	remaining: 26s
652:	learn: 0.0073333	total: 48.7s	remaining: 25.9s
653:	learn: 0.0073333	total: 48.8s	remaining: 25.8s
654:	learn: 0.0073333	total: 48.9s	remaining: 25.8s
655:	learn: 0.0073333	total: 49s	remaining: 25.7s
656:	learn: 0.0073333	total: 49s	remaining: 25.6s
657:	learn: 0.0073333	total: 49.1s	remaining: 25.5s
658:	learn: 0.0073333	total: 49.2s	remaining: 25.4s
659:	learn: 0.0073333	

799:	learn: 0.0073333	total: 59.2s	remaining: 14.8s
800:	learn: 0.0073333	total: 59.3s	remaining: 14.7s
801:	learn: 0.0073333	total: 59.3s	remaining: 14.7s
802:	learn: 0.0073333	total: 59.4s	remaining: 14.6s
803:	learn: 0.0073333	total: 59.5s	remaining: 14.5s
804:	learn: 0.0073333	total: 59.6s	remaining: 14.4s
805:	learn: 0.0073333	total: 59.6s	remaining: 14.4s
806:	learn: 0.0073333	total: 59.7s	remaining: 14.3s
807:	learn: 0.0073333	total: 59.8s	remaining: 14.2s
808:	learn: 0.0073333	total: 59.8s	remaining: 14.1s
809:	learn: 0.0073333	total: 59.9s	remaining: 14.1s
810:	learn: 0.0073333	total: 60s	remaining: 14s
811:	learn: 0.0073333	total: 1m	remaining: 13.9s
812:	learn: 0.0073333	total: 1m	remaining: 13.8s
813:	learn: 0.0073333	total: 1m	remaining: 13.8s
814:	learn: 0.0073333	total: 1m	remaining: 13.7s
815:	learn: 0.0073333	total: 1m	remaining: 13.6s
816:	learn: 0.0073333	total: 1m	remaining: 13.5s
817:	learn: 0.0073333	total: 1m	remaining: 13.5s
818:	learn: 0.0073333	total: 1m	remai

958:	learn: 0.0073333	total: 1m 10s	remaining: 3.02s
959:	learn: 0.0073333	total: 1m 10s	remaining: 2.95s
960:	learn: 0.0073333	total: 1m 10s	remaining: 2.88s
961:	learn: 0.0073333	total: 1m 10s	remaining: 2.8s
962:	learn: 0.0073333	total: 1m 11s	remaining: 2.73s
963:	learn: 0.0073333	total: 1m 11s	remaining: 2.65s
964:	learn: 0.0073333	total: 1m 11s	remaining: 2.58s
965:	learn: 0.0073333	total: 1m 11s	remaining: 2.51s
966:	learn: 0.0073333	total: 1m 11s	remaining: 2.43s
967:	learn: 0.0073333	total: 1m 11s	remaining: 2.36s
968:	learn: 0.0073333	total: 1m 11s	remaining: 2.29s
969:	learn: 0.0073333	total: 1m 11s	remaining: 2.21s
970:	learn: 0.0073333	total: 1m 11s	remaining: 2.14s
971:	learn: 0.0073333	total: 1m 11s	remaining: 2.06s
972:	learn: 0.0073333	total: 1m 11s	remaining: 1.99s
973:	learn: 0.0073333	total: 1m 11s	remaining: 1.92s
974:	learn: 0.0073333	total: 1m 11s	remaining: 1.84s
975:	learn: 0.0073333	total: 1m 11s	remaining: 1.77s
976:	learn: 0.0073333	total: 1m 12s	remaining: 

In [89]:
result8 = cat2.predict(X_val01)
pd.Series(result8).value_counts()

neg    10944
pos      456
dtype: int64

In [90]:
print('f1 score:', f1_score(y_val.replace({'neg': 0, 'pos' : 1}), pd.Series(result8).replace({'neg': 0, 'pos' : 1})))
print('accuracy score:', accuracy_score(y_val.replace({'neg': 0, 'pos' : 1}), pd.Series(result8).replace({'neg': 0, 'pos' : 1})))

f1 score: 0.5949177877428998
accuracy score: 0.9762280701754386


In [91]:
df2 = pd.DataFrame({'predicted':result8, 'true':y_val})
df2.reset_index(drop=True,inplace=True)

In [92]:
i = 0
j= 0
false_neg = 0
false_pos = 0

for predicted, true in df2.values:
    if predicted != true: #예측이 틀렸을 때
        if true == 'neg': 
            i = i+10
            false_pos = false_pos+1
        else :
            j = j+500
            false_neg = false_neg+1
            
print('CatBoost 총 비용 :','$',i+j)
print('positive를 negative로 분류 :',false_neg,'개', '$',false_neg*500)
print('negative를 positive로 분류 :',false_pos,'개', '$', false_pos*10)

CatBoost 총 비용 : $ 9570
positive를 negative로 분류 : 14 개 $ 7000
negative를 positive로 분류 : 257 개 $ 2570


### TESTing the Tuned Models

**1) XGB**

In [93]:
test_pred7 = xgb_model_2.predict(test1)
pd.Series(test_pred7).value_counts()

neg    15629
pos      371
dtype: int64

In [94]:
print('f1 score:', f1_score(test['class'].replace({'neg': 0, 'pos' : 1}), pd.Series(test_pred7).replace({'neg': 0, 'pos' : 1})))
print('accuracy score:', accuracy_score(test['class'].replace({'neg': 0, 'pos' : 1}), pd.Series(test_pred7).replace({'neg': 0, 'pos' : 1})))

f1 score: 0.9463806970509383
accuracy score: 0.9975


In [95]:
t_df7 = pd.DataFrame({'predicted':test_pred7, 'true':test['class']})
t_df7.reset_index(drop=True,inplace=True)

In [96]:
i = 0
j= 0
false_neg = 0
false_pos = 0

for predicted, true in t_df7.values:
    if predicted != true: #예측이 틀렸을 때
        if true == 'neg': 
            i = i+10
            false_pos = false_pos+1
        else :
            j = j+500
            false_neg = false_neg+1
            
print('XG Boost 총 비용 :','$',i+j)
print('positive를 negative로 분류 :', '$',false_neg,'개',false_neg*500)
print('negative를 positive로 분류 :', '$',false_pos,'개', false_pos*10)

XG Boost 총 비용 : $ 11180
positive를 negative로 분류 : $ 22 개 11000
negative를 positive로 분류 : $ 18 개 180


**1-1) XGBoost with scale_pos_weight = 52**<br>

In [97]:
test_pred10 = xgb_model_3.predict(test1)
pd.Series(test_pred10).value_counts()

neg    15583
pos      417
dtype: int64

In [98]:
print('f1 score:', f1_score(test['class'].replace({'neg': 0, 'pos' : 1}), pd.Series(test_pred10).replace({'neg': 0, 'pos' : 1})))
print('accuracy score:', accuracy_score(test['class'].replace({'neg': 0, 'pos' : 1}), pd.Series(test_pred10).replace({'neg': 0, 'pos' : 1})))

f1 score: 0.9141414141414141
accuracy score: 0.99575


In [99]:
t_df10 = pd.DataFrame({'predicted':test_pred10, 'true':test['class']})
t_df10.reset_index(drop=True,inplace=True)

In [100]:
i = 0
j= 0
false_neg = 0
false_pos = 0

for predicted, true in t_df10.values:
    if predicted != true: #예측이 틀렸을 때
        if true == 'neg': 
            i = i+10
            false_pos = false_pos+1
        else :
            j = j+500
            false_neg = false_neg+1
            
print('XG Boost 총 비용 :','$',i+j)
print('positive를 negative로 분류 :', '$',false_neg,'개',false_neg*500)
print('negative를 positive로 분류 :', '$',false_pos,'개', false_pos*10)

XG Boost 총 비용 : $ 7050
positive를 negative로 분류 : $ 13 개 6500
negative를 positive로 분류 : $ 55 개 550


**2) CatBoost with learning rate = 0.3, scale_pos_weight=10**

In [101]:
test_pred8 = cat2.predict(test1)
pd.Series(test_pred8).value_counts()

neg    15337
pos      663
dtype: int64

In [102]:
print('f1 score:', f1_score(test['class'].replace({'neg': 0, 'pos' : 1}), pd.Series(test_pred8).replace({'neg': 0, 'pos' : 1})))
print('accuracy score:', accuracy_score(test['class'].replace({'neg': 0, 'pos' : 1}), pd.Series(test_pred8).replace({'neg': 0, 'pos' : 1})))

f1 score: 0.7090558766859344
accuracy score: 0.981125


In [103]:
t_df8 = pd.DataFrame({'predicted':test_pred8, 'true':test['class']})
t_df8.reset_index(drop=True,inplace=True)

In [104]:
i = 0
j= 0
false_neg = 0
false_pos = 0

for predicted, true in t_df8.values:
    if predicted != true: #예측이 틀렸을 때
        if true == 'neg': 
            i = i+10
            false_pos = false_pos+1
        else :
            j = j+500
            false_neg = false_neg+1
            
print('CatBoost 총 비용 :','$',i+j)
print('positive를 negative로 분류 :',false_neg, '개', '$',false_neg*500)
print('negative를 positive로 분류 :',false_pos,'개', '$', false_pos*10)

CatBoost 총 비용 : $ 6450
positive를 negative로 분류 : 7 개 $ 3500
negative를 positive로 분류 : 295 개 $ 2950


## 분석과정에 대한 간단한 설명

#### 전처리:
column을 삭제할 경우, 정보 손실이 우려되어 na를 -1로 바꿔주는 방식으로 결측치를 처리하였다.

#### 데이터:
주어진 데이터를 training-validation set으로 나누어 학습시켰고, test set은 따로 Kaggle에서 가져온 것이다.  
training data 와 test set은 둘 다 imbalanced data로, training dataset은 약 pos:neg = 1:50의 비율이고 test dataset은 약 pos:neg= 1:40의 비율로 구성된다.  
주어진 원시 데이터는 stratify를 통해 각각의 클래스가 1:50=pos:neg에 맞게, 80:20 비율로 training과 validation data 로 나뉘었다.  
후에 training data를 SMOTE을 이용하여 oversampling하였다.  
즉, minority class였던 pos가 neg와 같은 수로 맞춰진 상태에서 모델이 정답 레이블을 학습한 것이다.  

#### 진행과정:
Random Forest, Catboost, XGBoost, Logistic Regression, Softmax Regression, SVC 모델을 각각 비교해보았다. 추가적으로 Catboost와 XGBoost의 경우에는, 파라미터를 조정해보며 더 우수한 결과를 도출할 수 있도록 하였다.

#### 결과:
파라미터를 조정한 Catboost가 가장 성능이 좋았다. 다음은 각 모델의 결과를 정리한 표이다.  

![title](cost_table.jpg)

## Hyperparameter Tuning 하며 알게된 것들 

1. 모델의 성능을 측정하는 기준인 f1 score은 높게, 우리 모델의 목적인 비용은 낮게 만드는 것은 쉽지 않은 일이다.<br>
대회의 목적은 총 비용을 줄이는 것이기 때문에, f1 score은 낮게 나오더라도 비용이 적게 나오는 hyperparameter을 썼다. <br>
false_negative의 페널티가 500이고, false_positive의 페널티가 10이다. 50배나 차이가 난다. false_positive가 100개 더 많아지는 것과 false_negative가 2개 많아지는 것이 똑같기 때문에, false_negative를 하나라도 더 줄이는 것에 집중했다. <br>
다만, validation set의 f1 score이 최대한 0.5 밑으로 떨어지는 것은 방지하려고 했다. 

2. validation set에서의 f1 score와 비용을 보고 파라미터를 조정했다. (주로 learning_rate 과 scale_pos_weight을 튜닝하였다)
<br>
그 후 Kaggle에서 가져온 test set을 모델에 넣어 결과를 도출했다. <br>

3. xgbClassifier의 최종 파라미터는 (learning_rate=0.2, subsample=0.8, objective="binary:logistic", scale_pos_weight=52, random_state=42)이다. <br>
다양하게 learning_rate과 scale_pos_weight 을 튜닝하며 결과를 지켜보았다.<br>
scale_pos_weight을 더 집중된 tuning이다. 52는 $\frac{sum-negative}{sum-positive}$ 이 공식에서 도출했다.
<br>

|  scale_pos_weight | 1 | 10 | 20 | 50  | 52  | 52  | 70  |
|---|---|---|---|---| --- | --- | --- |
|learning rate| 0.25 | 0.2 |  0.2 | 0.2 | 0.2 | 0.07 | 0.2 |
| **test 총 비용** | 11180 | 7890 | 7370 | 7550 | **7050**  | 7430  | 7610 |
|  false neg | 22 | 15 | 14 |14  |  13 | 14  |  14 |
| false pos | 18 | 39|  37| 55 | 55 | 43 | 61 |
<br>

사실 위의 공식에 따른 52를 scale_pos_weight에 넣는 것은 오버샘플링을 통해 클래스 수가 1:1로 맞춰진 학습 데이터를 생각한다면 적절치 않으나,<br>
learning rate을 0.2로 동일하게 놓고 scale_pos_weight을 각각 1과 52로 설정한 후 모델을 돌렸을 때, 총 비용 측면에서 52를 넣는 것이 결과가 더 좋았다. <br>
아마도 test dataset가 학습 데이터만큼 imbalanced data이기 때문에 모델과 fit이 좋지 않았나 생각한다.<br>
52가 모델에 무슨 의미를 지니는지 정확히 모르지만, scale_pos_weight에 (동일한 learning rate에서) 40, 50, 60 등을 추가로 넣어 돌렸을 때 52가 성능이 가장 우수하여 이 수치를 선택하였다.

4. CatBoostClassifier의 최종 파라미터는 cat2 = CatBoostClassifier(learning_rate=0.3, scale_pos_weight=10, verbose=True) 이다. <br>
다양한 learning rate과 scale_pos_weight을 튜닝하며 결과를 내보았다.

| learning rate |  0.1 | 0.1  | 0.15  | 0.2  |  0.2  | 0.3  |
|---|---|---|---|---|---|---|
|scale_pos_weight| 5 | 20 | 10 | 1 | 10 | 10 |
| **총 비용**  | 8410  |  8090 |  6910 | 11810  | 7120  | **5920**  |
| false neg  |  12 | 5  |  7 |  23 |  7 |  6 |
| false pos |  241 |  559 | 341  | 31  | 362   | 292  |